Extract leaderboard fetch logic from song select beatmap leaderboard drawable

bdach opened the pull request:
RFC. Another attempt at this.

- Supersedes https://github.com/ppy/osu/pull/31881
- Supersedes / closes https://github.com/ppy/osu/pull/31355
- Closes https://github.com/ppy/osu/issues/29861
- Closes https://github.com/ppy/osu/issues/26716
- Closes https://github.com/ppy/osu/issues/27609

This is a weird diff because I am feeling rather boxed in by all the constraints, namely that:

- Leaderboard state should be global state
- But the global state is essentially managed by song select and namely `BeatmapLeaderboard` itself. That's because trying to e.g. not have `BeatmapLeaderboard` pass the beatmap and the ruleset to the global leaderboard manager is worse, as it essentially introduces two parallel paths of execution that need to be somehow merged into one (as in I'd have to somehow sync `LeaderboardManager` responding to global beatmap/ruleset changes with `BeatmapLeaderboard` which is also doing the same via `RefetchScores()` inheritance hell)
- Also local leaderboard fetching is data-push (as in the scores can change under the leaderboard manager), and online leaderboard fetching is data-pull (as in the scores do not change unless the leaderboard manager does something). Also online leaderboard fetching can fail. Which is why I need to still have the weird setup wherein there's a `FetchWithCriteriaAsync()` (because I need to be able to respond to online requests taking time, or failing), but also the `BeatmapLeaderboard` only uses the public `Scores` bindable to actually read the scores (because it needs to respond to new local scores arriving).
- Another thing to think about here is what happens when a retrieval fails because e.g. the user requested friend leaderboards without having supporter. With how this diff is written, that special condition is handled local to `BeatmapLeaderboard`, and `LeaderboardManager`'s state will remain as whatever it was before that scope change was requested, which may be considered good or it may not (I imagine it's better to show any scores in gameplay than not in this case, but maybe I'm wrong?)

bdach committed:
Extract leaderboard fetch logic from song select beatmap leaderboard drawable

RFC. Another attempt at this.

- Supersedes https://github.com/ppy/osu/pull/31881
- Supersedes / closes https://github.com/ppy/osu/pull/31355
- Closes https://github.com/ppy/osu/issues/29861

This is a weird diff because I am feeling rather boxed in by all the
constraints, namely that:

- Leaderboard state should be global state
- But the global state is essentially managed by song select and namely
  `BeatmapLeaderboard` itself. That's because trying to e.g. not have
  `BeatmapLeaderboard` pass the beatmap and the ruleset to the global
  leaderboard manager is worse, as it essentially introduces two
  parallel paths of execution that need to be somehow merged into one
  (as in I'd have to somehow sync `LeaderboardManager` responding to
  beatmap/ruleset changes with `BeatmapLeaderboard` which is inheritance
  hell)
- Also local leaderboard fetching is data-push (as in the scores can
  change under the leaderboard manager), and online leaderboard fetching
  is data-pull (as in the scores do not change unless the leaderboard
  manager does something). Also online leaderboard fetching can fail.
  Which is why I need to still have the weird setup wherein there's a
  `FetchWithCriteriaAsync()` (because I need to be able to respond to
  online requests taking time, or failing), but also the
  `BeatmapLeaderboard` only uses the public `Scores` bindable to
  actually read the scores (because it needs to respond to new local
  scores arriving).
- Another thing to think about here is what happens when a retrieval
  fails because e.g. the user requested friend leaderboards without
  having supporter. With how this diff is written, that special
  condition is handled to `BeatmapLeaderboard`, and
  `LeaderboardManager`'s state will remain as whatever it was before
  that scope change was requested, which may be considered good or it
  may not (I imagine it's better to show scores in gameplay than not in
  this case, but maybe I'm wrong?)

smoogipoo commented:
This looks better to me on a brief pass. I'll defer full review to @peppy.

peppy commented:
Sorry for the late response. I am also on board with this direction. It seems that this is a soft dependency for moving forward with song select, so would be interested in seeing things move forward so it's not blocking (although keep in mind there's still *a lot* of [other things](https://github.com/ppy/osu/issues/32736) blocking release so it's not super urgent).

peppy reviewed (commented):
As commented.

bdach committed:
Merge branch 'master' into yet-another-leaderboard-rewrite-attempt

bdach committed:
Fix test failures

bdach committed:
Simplify code further

bdach committed:
Fix various weirdness around refetching

bdach committed:
Actually use leaderboard manager in players

This was supposed to be the case all along, but I guess in all of the
rewrite attempts I forgot?

With this, https://github.com/ppy/osu/issues/29861 is actually fixed
again - and additionally, so is https://github.com/ppy/osu/issues/26716.

bdach commented:
Alright, well, I fixed test failures as of https://github.com/ppy/osu/pull/32494/commits/0209129618e63f5ca6d22bc9dd283f7735fb3466, also fixed some more breakage that only showed in-game, and also actually changed this so that the players use the new global state in https://github.com/ppy/osu/pull/32494/commits/16b817be1d91169f5c0125de6f1656afc1faf9bb, which fixes like three separate issues.

Should be ready for proper review. I'll check back on CI later.

bdach committed:
Update scores request to move away from old endpoint

bdach commented:
As I typed my previous reply I realised there is probably an issue here with the replay player usage. Blocking temporarily until confirmation.

bdach committed:
Ensure correct global leaderboard state when presenting scores

With `ReplayPlayer` now consuming the `LeaderboardManager`'s global
state, flows such as presenting a score need to set the global state up
correctly to avoid accidentally showing a leaderboard from a completely
different score.

This also incidentally closes https://github.com/ppy/osu/issues/27609.

bdach commented:
Issue in question should be resolved by https://github.com/ppy/osu/pull/32494/commits/31b98ac7b99414870c6b02fb472d5ab39d64fa44. See commit message for details what the issue was.
